app/code/Magento/InventoryLowQuantityNotification/Model/ResourceModel/BulkConfigurationTransfer.php

```diff
@@ -65,37 +65,35 @@ public function execute(
 
         $types = $this->getProductTypesBySkus->execute($skus);
 
-        foreach ($types as $sku => $type) {
-            if ($this->isSourceItemManagementAllowedForProductType->execute($type)) {
-                foreach ($skus as $sku) {
-                    $qry = $connection
-                        ->select()
-                        ->from($tableName, 'notify_stock_qty')
-                        ->where('sku = ?', $sku)
-                        ->where('source_code = ?', $originSource);
+        foreach ($skus as $sku) {
+            if ($this->isSourceItemManagementAllowedForProductType->execute($types[$sku])) {
+                $qry = $connection
+                    ->select()
+                    ->from($tableName, 'notify_stock_qty')
+                    ->where('sku = ?', $sku)
+                    ->where('source_code = ?', $originSource);
 
-                    $res = $connection->fetchOne($qry);
+                $res = $connection->fetchOne($qry);
 
-                    $notifyStockQty = ($res === null || $res === false) ? null : (float) $res;
-                    try {
-                        $connection->insert(
+                $notifyStockQty = ($res === null || $res === false) ? null : (float) $res;
+                try {
+                    $connection->insert(
+                        $tableName,
+                        [
+                            'source_code' => $destinationSource,
+                            'sku' => $sku,
+                            'notify_stock_qty' => $notifyStockQty,
+                        ]
+                    );
+                } catch (DuplicateException $e) {
+                    // Do not overwrite an existing configuration if the item was not assigned to the source
+                    if ($res !== false) {
+                        $connection->update(
                             $tableName,
-                            [
-                                'source_code' => $destinationSource,
-                                'sku' => $sku,
-                                'notify_stock_qty' => $notifyStockQty,
-                            ]
+                            ['notify_stock_qty' => $notifyStockQty],
+                            $connection->quoteInto('sku = ?', $sku) . ' AND ' .
+                            $connection->quoteInto('source_code = ?', $destinationSource)
```

Comment:
This logic looks like implementation on the app level of INSERT ON DUPLICATE UPDATE  MySQL query.

It's more efficient to do that on DB level